### PR TITLE
Merge the concepts branch

### PR DESCRIPTION
**Pull request by zah**
As evident from the new test cases and the closed issues, the current state of the concepts branch is quite much better than what "devel" supports.

Until this branch is merged, I'll be always hesitant to look at issues involving generics, because all possible fixes have a high potential for conflicts with this branch. Now that I'm contributing regularly again, I think I'll be able to finish the rest of the features advertised in the manual in the next few weeks.

### CODE REVIEW

**Comment by Araq** [comment]
This breaks https://github.com/unicredit/emmy which uses concepts heavily.

**Comment by andreaferretti** [comment]
I'll be happy to adapt Emmy to make it work with the new branch, if it brings more stable concepts. Right now, Emmy is a little stuck waiting for concepts to improve

**Comment by zah** [comment]
@andreaferretti, I found the issue in building emmy.

The new concept branch tries to be backwards-compatible with the old code, but it does introduce some new features that can improve the style of your concepts. Actually, I was inspired by emmy and included some examples from it in the manual (following the new recommended style):

https://github.com/nim-lang/Nim/blob/01bc5f32d665ea1c72590609391769cef60d6e83/doc/manual/generics.txt#L263

**Comment by vegansk** [comment]
@zah, is there any way to describe the Functor using this branch? Something like this (not working):
```nim
import future

type
  Functor[T] = concept x
    type A = auto
    map(x, T -> A) == Functor[A]

type F[T] = object
  v: T

proc map[A,B](v: F[A], f: A -> B): F[B] =
  F(v: f(v.v))

echo F is Functor
```

**Comment by zah** [comment]
@vegansk, I've documented one potentially confusing aspect of the generic concepts here:

https://github.com/nim-lang/Nim/blob/01207b6cfda148c513d5b3fe7db024d1b5881a95/doc/manual/generics.txt#L436

I've used the Functor as an example in the manual. Alas, it still doesn't work, but this is caused by another issue that just needs to be fixed.

**Comment by zah** [comment]
The failed test case above seems like a false-positive:

```
FAIL: tioselectors.nim
Test "tests/async\tioselectors.nim" in category "async"
Failure: reExitcodesDiffer
Expected:
exitcode: 0
Gotten:
exitcode: 1
Output:
Traceback (most recent call last)
tioselectors.nim(614)    tioselectors
tioselectors.nim(513)    socket_notification_test
system.nim(3537)         failedAssertImpl
system.nim(3529)         raiseAssert
system.nim(2639)         sysFatal
Error: unhandled exception: len(rcm) == 2  [AssertionError]
PASS: tioselectors.nim -d:upcoming
```
I haven't looked deeply into it, but perhaps the test depends on a working network.

File: doc/manual/generics.txt [hunk 1]
```diff
@@ -425,10 +429,25 @@ types, thus allowing you to encode implementation detail types:
...
-Enumerable[T] just by their short name. Much like the regular generic types,
+`Enumerable[T]` just by their short name. Much like the regular generic types,
 the concept will be automatically instantiated with the bind once auto type
 in the place of each missing generic param.
 
+Please note that generic concepts such as `Enumerable[T]` can be matched
+against concrete types such as `string`. Nim doesn't require the concept
+type to have the same number of parameters as the type being matched.
+In order to express such a requirement, you'll need to rely on a type
+mapping operator such a `genericHead` or `stripGenericParams` within the
```
Comment (Araq) [inline doc/manual/generics.txt]: 'such as'

File: doc/manual/generics.txt [hunk 2]
```diff
@@ -425,10 +429,25 @@ types, thus allowing you to encode implementation detail types:
...
+concept body:
+
+.. code-block:: nim
+  import future, typetraits
+
+  type
+    Functor[A] = concept f
+      f.value is A
+      map(f, A -> T1) is genericHead(f.type)[T1]
```
Comment (Araq) [inline doc/manual/generics.txt]: This is incomprehensible. It needs at least a concrete implementation that matches and an explanation of why it matches.

File: lib/system.nim [hunk 3]
```diff
@@ -98,6 +98,17 @@ type
   SomeNumber* = SomeInteger|SomeReal
     ## type class matching all number types
 
+  T1* = auto
```
Comment (Araq) [inline lib/system.nim]: Can't these be in the ``typetraits`` module instead?

File: compiler/sigmatch.nim [hunk 4]
```diff
@@ -1363,6 +1398,9 @@ proc paramTypesMatchAux(m: var TCandidate, f, argType: PType,
     argType = argType
     c = m.c
 
+  if inferTypeClassParam(c, f, argType):
```
Comment (Araq) [inline compiler/sigmatch.nim]: Move ``c.inTypeClass == 0`` check to here for better performance.

File: compiler/semdata.nim [hunk 5]
```diff
@@ -238,8 +249,11 @@ proc makeVarType*(c: PContext, baseType: PType): PType =
     addSonSkipIntLit(result, baseType.assertNotNil)
 
 proc makeTypeDesc*(c: PContext, typ: PType): PType =
-  result = newTypeS(tyTypeDesc, c)
-  result.addSonSkipIntLit(typ.assertNotNil)
+  if typ.kind == tyTypeDesc:
```
Comment (Araq) [inline compiler/semdata.nim]: Last time I thought about it, typedesc[typedesc[T]]  is not the same as typedesc[T] and conflating these is not sound.

File: compiler/ast.nim [hunk 6]
```diff
@@ -1279,6 +1289,8 @@ proc copyType*(t: PType, owner: PSym, keepId: bool): PType =
     when debugIds: registerId(result)
   result.sym = t.sym          # backend-info should not be copied
 
+proc exactReplica*(t: PType): PType = copyType(t, t.owner, true)
```
Comment (Araq) [inline compiler/ast.nim]: General remark: This is a bad idea. Types can be mutated in the codegen ``field.loc`` etc and the write only affects one "replica" and not the other, causing subtle codegen bugs.

File: compiler/sigmatch.nim [hunk 7]
```diff
@@ -718,8 +722,51 @@ proc typeRel(c: var TCandidate, f, aOrig: PType, doBind = true): TTypeRelation =
...
+    aOrig = if useTypeLoweringRuleInTypeClass:
+          aOrig.skipTypes({tyTypeDesc, tyFieldAccessor})
+        else:
+          aOrig
+
+  if aOrig.kind == tyInferred:
+    # echo "INFER A"
+    # debug f
+    # debug aOrig
+    let prev = aOrig.previouslyInferred
```
Comment (Araq) [inline compiler/sigmatch.nim]: This section should be put into a helper proc, maybe even .noinline. It's a cold code path.

File: compiler/ast.nim [hunk 8]
```diff
@@ -354,44 +354,52 @@ type
...
 
-    tyAnd, tyOr, tyNot #\
+    tyInferred
+      # In the initial state `base` stores a type class constraining
+      # the types that can be inferred. After a candidate type is
+      # selected, it's stored in `lastSon`. Between `base` and `lastSon`
+      # there may be 0, 2 or more types that were also considered as
+      # possible candidates in the inference process (i.e. lastSon will
+      # be updated to store a type best conforming to all candidates)
+
```
Comment (Araq) [inline compiler/ast.nim]: These metatypes always creep into every place in the compiler. Did you try to use the tyAlias mechanism for this instead?

Comment (zah) [inline compiler/ast.nim]: I plan to remove `tyInferred` in one of the next commits. It was born as an optimization that allowed me to get away without using type binding tables (as in sigmatch), but some other aspects of the implementation made me reconsider this decision. Please don't make me remove it before the merge :)

File: compiler/semtypes.nim [hunk 9]
```diff
@@ -845,9 +845,9 @@ proc liftParamType(c: PContext, procKind: TSymKind, genericParams: PNode,
 
     for i in 0 .. paramType.sonsLen - 2:
       if paramType.sons[i].kind == tyStatic:
-        var x = copyNode(ast.emptyNode)
-        x.typ = paramType.sons[i]
-        result.rawAddSon makeTypeFromExpr(c, x) # aka 'tyUnknown'
+        var staticCopy = paramType.sons[i].exactReplica
```
Comment (Araq) [inline compiler/semtypes.nim]: maybe add an assertion to exactReplica so that it's only used for tyStatic et al where it cannot do harm for the backends?

File: compiler/ast.nim (hunk 6 above)
Comment (zah) [inline compiler/ast.nim]: I think this code pre-dates the introduction of `tyAlias` (or at least my familiarity with it). I'll consider using it in the future. For now, I think our time is better spent implementing the rest of the spec in the manual and fixing all concept-related issues in the tracker.

File: lib/system.nim (hunk 3 above)
Comment (zah) [inline lib/system.nim]: These helpers are useful for writing implicitly generic procs as in:

``` Nim
proc find(s: seq[T1], value: T1)
```

We've discussed them a bit on IRC in the past and I managed to convince you back then that they belong to the system module. In general, the are not related to `typetraits` in any way.

File: doc/manual/generics.txt (hunk 2 above)
Comment (zah) [inline doc/manual/generics.txt]: I'll expand this or come up with a simpler example. `Functor` is a well-known concept from Haskell, and that's why I considered it a clear-enough example.

File: compiler/semtypes.nim [hunk 10]
```diff
@@ -1226,6 +1226,13 @@ proc fixupTypeOf(c: PContext, prev: PType, typExpr: PNode) =
     result.sym = prev.sym
     assignType(prev, result)
 
+proc symFromExpectedTypeNode(c: PContext, n: PNode): PSym =
+  if n.kind == nkType:
+    result = symFromType(n.typ, n.info)
+  else:
+    localError(n.info, "xx")
```
Comment (Araq) [inline compiler/semtypes.nim]: I hope this is not in the final version.

File: compiler/semstmts.nim [hunk 11]
```diff
@@ -1553,6 +1553,11 @@ proc usesResult(n: PNode): bool =
       for c in n:
         if usesResult(c): return true
 
+proc inferConceptStaticParam(c: PContext, typ: PType, n: PNode) =
+  let res = semConstExpr(c, n)
+  if not sameType(res.typ, typ.base): localError(n.info, "")
```
Comment (Araq) [inline compiler/semstmts.nim]: no error message.

File: compiler/semstmts.nim [hunk 12]
```diff
@@ -1604,10 +1609,19 @@ proc semStmtList(c: PContext, n: PNode, flags: TExprFlags): PNode =
...
     else:
-      n.sons[i] = semExpr(c, n.sons[i])
-      if c.inTypeClass > 0 and n[i].typ != nil:
-        case n[i].typ.kind
+      var expr = semExpr(c, n.sons[i])
+      n.sons[i] = expr
+      if c.inTypeClass > 0 and expr.typ != nil:
+        case expr.typ.kind
         of tyBool:
+          if expr.kind == nkInfix and expr[0].sym.name.s == "==":
```
Comment (Araq) [inline compiler/semstmts.nim]: ``and expr[0].kind == nkSym``

File: compiler/msgs.nim [hunk 13]
```diff
@@ -929,7 +927,7 @@ proc rawMessage*(msg: TMsgKind, args: openArray[string]) =
     if msg notin gNotes: return
     title = HintTitle
     color = HintColor
-    kind = HintsToStr[ord(msg) - ord(hintMin)]
+    if msg != hintUserRaw: kind = HintsToStr[ord(msg) - ord(hintMin)]
```
Comment (Araq) [inline compiler/msgs.nim]: why is this check necessary?

File: compiler/parser.nim [hunk 14]
```diff
@@ -770,6 +771,13 @@ proc parseOperators(p: var TParser, headNode: PNode,
 
 proc simpleExprAux(p: var TParser, limit: int, mode: TPrimaryMode): PNode =
   result = primary(p, mode)
+  if p.tok.tokType == tkCurlyDotLe and
+     p.lex.lineNumber == result.info.line and
```
Comment (Araq) [inline compiler/parser.nim]: Will check, but I doubt this rule can be specified in the (already convoluted) grammar.

File: compiler/semcall.nim [hunk 15]
```diff
@@ -212,7 +210,6 @@ proc resolveOverloads(c: PContext, n, orig: PNode,
       if result.state != csMatch:
         n.sons.delete(1)
         orig.sons.delete(1)
-        excl n.flags, nfExprCall
```
Comment (Araq) [inline compiler/semcall.nim]: Why is this removed? It's a bugfix!

File: compiler/semcall.nim [hunk 16]
```diff
@@ -367,31 +358,48 @@ proc semResolvedCall(c: PContext, n: PNode, x: TCandidate): PNode =
 
 proc canDeref(n: PNode): bool {.inline.} =
   result = n.len >= 2 and (let t = n[1].typ;
-    t != nil and t.skipTypes({tyGenericInst, tyAlias}).kind in {tyPtr, tyRef})
```
Comment (Araq) [inline compiler/semcall.nim]: Why is 'tyAlias' removed here?

File: compiler/semcall.nim [hunk 17]
```diff
@@ -367,31 +358,48 @@ proc semResolvedCall(c: PContext, n: PNode, x: TCandidate): PNode =
...
+    var r = resolveOverloads(c, n, nOrig, filter, flags, errors)
     if r.state == csMatch: result = semResolvedCall(c, n, r)
     else:
       # get rid of the deref again for a better error message:
       n.sons[1] = n.sons[1].sons[0]
       notFoundError(c, n, errors)
   else:
-    notFoundError(c, n, errors)
-  # else: result = errorNode(c, n)
+    if efExplain notin flags:
```
Comment (Araq) [inline compiler/semcall.nim]: So ... independently of whether there is a concept involved, always retry the overload resolution with enabling all the diagnostic output? This seems slow for ``when compiles`` context as well as nimsuggest.

File: compiler/semcall.nim [hunk 18]
```diff
@@ -460,7 +471,7 @@ proc searchForBorrowProc(c: PContext, startScope: PScope, fn: PSym): PSym =
   call.add(newIdentNode(fn.name, fn.info))
   for i in 1.. <fn.typ.n.len:
     let param = fn.typ.n.sons[i]
-    let t = skipTypes(param.typ, abstractVar-{tyTypeDesc, tyDistinct})
```
Comment (Araq) [inline compiler/semcall.nim]: why?

File: compiler/semcall.nim [hunk 19]
```diff
@@ -470,12 +481,6 @@ proc searchForBorrowProc(c: PContext, startScope: PScope, fn: PSym): PSym =
       x = t.baseOfDistinct
     call.add(newNodeIT(nkEmpty, fn.info, x))
   if hasDistinct:
-    var resolved = semOverloadedCall(c, call, call, {fn.kind})
+    var resolved = semOverloadedCall(c, call, call, {fn.kind}, {})
     if resolved != nil:
       result = resolved.sons[0].sym
-      if not compareTypes(result.typ.sons[0], fn.typ.sons[0], dcEqIgnoreDistinct):
```
Comment (Araq) [inline compiler/semcall.nim]: why? these are bugfixes.

File: compiler/sigmatch.nim [hunk 20]
```diff
@@ -22,7 +22,13 @@ type
   TCandidateState* = enum
     csEmpty, csMatch, csNoMatch
 
-  CandidateErrors* = seq[(PSym,int)]
+  CandidateError = tuple
```
Comment (Araq) [inline compiler/sigmatch.nim]: should be an object

File: compiler/sigmatch.nim [hunk 21]
```diff
@@ -577,14 +588,15 @@ proc typeRangeRel(f, a: PType): TTypeRelation {.noinline.} =
 
 proc matchUserTypeClass*(c: PContext, m: var TCandidate,
                          ff, a: PType): PType =
-  var body = ff.skipTypes({tyUserTypeClassInst})
+  var
+    Concept = ff.skipTypes({tyUserTypeClassInst})
```
Comment (Araq) [inline compiler/sigmatch.nim]: use 'concpt'

File: compiler/ast.nim [hunk 22]
```diff
@@ -1581,7 +1581,7 @@ proc hasPattern*(s: PSym): bool {.inline.} =
   result = isRoutine(s) and s.ast.sons[patternPos].kind != nkEmpty
 
 iterator items*(n: PNode): PNode =
-  for i in 0.. <n.len: yield n.sons[i]
+  for i in 0.. <n.safeLen: yield n.sons[i]
```
Comment (Araq) [inline compiler/ast.nim]: that should be a ``safeItems`` iterator then? I dislike code like this, the client should know enough about the node kind to at least be able to say whether it's an atomic node or not.

File: compiler/ast.nim [hunk 23]
```diff
@@ -494,6 +494,8 @@ type
     tfGenericTypeParam
     tfImplicitTypeParam
     tfInferrableStatic
+    tfExplicit        # for typedescs, marks types explicitly prefixed with the
```
Comment (Araq) [inline compiler/ast.nim]: So which part of the spec this deals with? I don't like it. :-) Why is ``type x`` suddenly different from e.g.  ``typeDesc[int]``?

Comment (zah) [inline compiler/ast.nim]: Within the concept body, types can appear in positions where ordinary values
and parameters are expected. This provides a more convenient way to check for
the presence of callable symbols with specific signatures:

``` nim
  type
    OutputStream = concept var s
      s.write(string)
```

In order to check for symbols accepting ``typedesc`` params, you must prefix
the type with an explicit ``type`` modifier. The named instance of the type,
following the ``concept`` keyword is also considered an explicit ``typedesc``
value that will be matched only as a type.

``` nim
  type
    # Let's imagine a user-defined casting framework with operators
    # such as `val.to(string)` and `val.to(JSonValue)`. We can test
    # for these with the following concept:
    MyCastables = concept x
      x.to(type string)
      x.to(type JSonValue)

    # Let's define a couple of concepts, known from Algebra:
    AdditiveMonoid* = concept x, y, type T
      x + y is T
      T.zero is T # require a proc such as `int.zero` or 'Position.zero'

    AdditiveGroup* = concept x, y, type T
      x is AdditiveMonoid
      -x is T
      x - y is T
```

File: compiler/semdata.nim (hunk 5 above)
Comment (zah) [inline compiler/semdata.nim]: Supporting these higher-order types is a display of strength, but I'm afraid the rest of the compiler is not there yet. There are a lot of places where various flags and conditions have to be checked more carefully. It's a worthy goal for a future refactoring.

File: compiler/sigmatch.nim (hunk 7 above)
Comment (zah) [inline compiler/sigmatch.nim]: I'll address this in the future when `tyInferred` gets removed.

File: compiler/msgs.nim (hunk 13 above)
Comment (zah) [inline compiler/msgs.nim]: `hintUserRaw` is a the very last element in the Hints sequence. It's used when you don't want to output even the "Hint: " qualification, but you still want to collect the message as a diagnostic. I used in one place to list the overloads that were considered while trying to match a concept.

File: compiler/parser.nim (hunk 14 above)
Comment (zah) [inline compiler/parser.nim]: The goal of this patch is to enable support for attaching pragmas to call sites. I think that's quite useful ability in general. In this branch, it's used for the "explain" pragma.

File: compiler/semcall.nim (hunk 15 above)
Comment (zah) [inline compiler/semcall.nim]: Hrm, the Github diff viewer is messing something up here. This line is not removed in the actual code on my hard drive.

File: compiler/semcall.nim (hunk 16 above)
Comment (zah) [inline compiler/semcall.nim]: Wrong diff again here. These lines display as "outdated", so perhaps there is some explanation for this behavior.

File: compiler/semcall.nim (hunk 17 above)
Comment (zah) [inline compiler/semcall.nim]: I think nimsuggest should display this diagnostic, but I'll add a check for "compiles".

File: compiler/semcall.nim (hunk 18 above)
Comment (zah) [inline compiler/semcall.nim]: Another case of invalid diff.

File: compiler/semcall.nim (hunk 19 above)
Comment (zah) [inline compiler/semcall.nim]: Another case of invalid diff.

File: compiler/semcall.nim (hunk 15 above)
Comment (dom96) [inline compiler/semcall.nim]: There is nothing wrong with GitHub, Araq has reviewed each commit individually so this diff is for one of your commits.

Comment (zah) [inline compiler/semcall.nim]: I got the impression that the "changed files" summary for the whole pull request also included that line, but now I see that's not the case. 

I now remember what happened. One of the merges was having conflicts all over the place in semcall up, so I stared from my own code and carefully migrated all the changes from the devel branch by hand (line by line):
https://github.com/nim-lang/Nim/pull/5606/commits/d6d548c32f4d60bd4f89c377fb94bfc57ad4cbfd

File: compiler/ast.nim (hunk 22 above)
Comment (zah) [inline compiler/ast.nim]: Unfortunately, I don't recall where this was needed.

File: compiler/parser.nim (hunk 14 above)
Comment (Araq) [inline compiler/parser.nim]: I know. But if you cannot write it down as a grammar rule, that's bad.
